integration/storagetest/logtests.go

```diff
@@ -465,49 +463,61 @@ func (*logTests) TestDequeueLeaves(ctx context.Context, t *testing.T, s storage.
 	}
 
 	// Now try to dequeue them
-	cctx, cancel := context.WithTimeout(ctx, 3*time.Second) // Retry until timeout
-	defer cancel()
-	if err := s.ReadWriteTransaction(cctx, tree,
-		func(ctx context.Context, tx2 storage.LogTreeTX) error {
-			leaves2, err := tx2.DequeueLeaves(ctx, 99, fakeDequeueCutoffTime)
-			if err != nil {
-				t.Fatalf("Failed to dequeue leaves: %v", err)
-			}
-			if len(leaves2) != leavesToInsert {
-				t.Logf("Dequeued %d leaves but expected to get %d", len(leaves2), leavesToInsert)
-				return status.Errorf(codes.Aborted, "retry")
-			}
-			ensureAllLeavesDistinct(leaves2, t)
-			iTimestamp := ptypes.TimestampNow()
-			for i, l := range leaves2 {
-				l.IntegrateTimestamp = iTimestamp
-				l.LeafIndex = int64(i)
-			}
-			if err := tx2.UpdateSequencedLeaves(ctx, leaves2); err != nil {
-				t.Fatalf("UpdateSequencedLeaves(): %v", err)
-			}
-			return nil
-		}); err != nil {
-		t.Fatalf("Could not dequeue the expected number of leaves: %v", err)
+	// Some dequeue implementations probabalistically dequeue and require retrying until timeout.
+	leaves2 := dequeueLeaves(ctx, t, s, tree, fakeDequeueCutoffTime, 5, 0)
+	if len(leaves2) != leavesToInsert {
+		t.Fatalf("Dequeued %d leaves but expected to get %d", len(leaves2), leavesToInsert)
 	}
 
 	// If we dequeue again then we should now get nothing
-	if err := s.ReadWriteTransaction(cctx, tree,
+	if err := s.ReadWriteTransaction(ctx, tree,
 		func(ctx context.Context, tx3 storage.LogTreeTX) error {
 			leaves3, err := tx3.DequeueLeaves(ctx, 99, fakeDequeueCutoffTime)
 			if err != nil {
 				t.Fatalf("Failed to dequeue leaves (second time): %v", err)
 			}
 			if len(leaves3) != 0 {
-				t.Logf("Dequeued %d leaves but expected to get %d", len(leaves3), leavesToInsert)
-				return status.Errorf(codes.Aborted, "retry")
+				t.Fatalf("Dequeued %d leaves but expected to get %d", len(leaves3), leavesToInsert)
 			}
 			return nil
 		}); err != nil {
 		t.Fatalf("Could not dequeue the expected number of leaves: %v", err)
 	}
 }
 
+// dequeueLeaves starts a transaction and uses it to dequeue leaves.
+func dequeueLeaves(ctx context.Context, t *testing.T, ls storage.LogStorage, tree *trillian.Tree, ts time.Time, limit int, startIndex int64) []*trillian.LogLeaf {
```

Comment:
Should this be renamed to `dequeueAndSequenceLeaves()` or similar?

Reply:
yes. done